2020/day8/solution.py

```diff
@@ -0,0 +1,75 @@
+import re
+import typing
+from collections.abc import Iterable, Callable
+
+#Type Information
+Instruction = tuple[str, int]
+
+instruction_re = re.compile(r"(?P<instruction>\w{3}) (?P<value>[+-]\d+)")
+
+class Machine:
+
+    def _acc(self, value : int) -> int:
+        self.acc += value
+        return 1
+
+    INSTRUCTION_TABLE : dict[str, Callable[['Machine',int], int]] = {
+        'acc' : _acc,
+        'jmp' : lambda self, x: x,
+        'nop' : lambda self, x: 1,
+    }
+
+    def __init__(self, instructions : Iterable[Instruction]):
+        self.instructions = list(instructions)
+        self.reset()
+
+    def execute(self) -> bool:
+        self.reset()
+        executed_instructions = set()
+        while self.ip not in executed_instructions:
+            executed_instructions.add(self.ip)
+            if self.ip >= len(self.instructions):
+                return True
+            self.ip += self.execute_instruction(self.ip)
+        return False
```

Comment:
It's a bit strange in a VM for the main condition to be `self.ip not in executed_instructions` - it seems like that's specific to this problem. Instead, it would be best to have the main loop condition be the `self.ip >= len(self.instructions)` check. Rewriting gives us

```
    def execute(self) -> bool:
        self.reset()
        executed_instructions = set()
        while self.ip < len(self.instructions):
            if self.ip in executed_instructions:
                return False
            executed_instructions.add(self.ip)
            self.ip += self.execute_instruction(self.ip)
        return True
```
(This also fixes a bad array index in case `self.instructions` was of size 0!)

Then, it's time to apply the unaging advice to split loops. An `execute` function for a VM should really focus on executing instructions; the checking for loops by keeping track of instruction pointers is fairly problem specific. Here's one way:

```
    def execute(self):  # type annotation left as an exercise to the reader :)
        self.reset()
        while self.ip < len(self.instructions):
            yield self.ip
            self.ip += self.execute_instruction(self.ip)

    def execute_without_looping(self) -> bool:
        executed_instructions = set()
        for ip in self.execute():
            if self.ip in executed_instructions:
                return False
        return True
```

Then you can use `execute_without_looping` in `find_accumulator_before_loop` and `repair`, and have `execute` ready for future problems.

Reply:
Oh, I really like splitting it this way.  I agree, the checking for a loop really shouldn't be in the core, but I didn't know a good way to track executed ips without doing it in the loop.  Thanks for this.  Let me play  with it in a rewrite to better understand.